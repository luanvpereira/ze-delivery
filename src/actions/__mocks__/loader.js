export const hideLoader = jest.fn().mockImplementation(() => ({
	type: Math.floor(Math.random() * 1000)
}));

export const showLoader = jest.fn().mockImplementation(() => ({
	type: Math.floor(Math.random() * 1000)
}));
