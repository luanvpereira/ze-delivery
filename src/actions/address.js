import axios from 'axios';

import { GEOCODE_URL, GOOGLE_CREDENTIAL } from '../constants';

export const getAddressess = address => async () => {
	const {
		data: { results: addresses }
	} = await axios.get(`${GEOCODE_URL}?address=${address}&key=${GOOGLE_CREDENTIAL}`);

	return addresses;
};
