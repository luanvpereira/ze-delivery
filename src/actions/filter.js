import {
	CHANGE_METHOD_ADDRESS,
	CHANGE_CURRENT_CATEGORY,
	CHANGE_FILTER_TEXT
} from '../types';

export const setCurrentMethod = payload => ({
	type: CHANGE_METHOD_ADDRESS,
	payload
});

export const changeCategory = payload => ({
	type: CHANGE_CURRENT_CATEGORY,
	payload: parseInt(payload, 10)
});

export const changeFilterText = payload => ({
	type: CHANGE_FILTER_TEXT,
	payload
});
