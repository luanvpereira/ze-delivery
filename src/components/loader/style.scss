.loader {
	position: fixed;
	top: 0;
	width: 100%;
	height: 100vh;
	background-color: rgba(0,0,0,.7);
	z-index: 2;


	&:before {
		content: '';
		display: block;
		width: 100px;
		height: 100px;
		background-image: url('../../lib/assets/pulse.svg');
		background-repeat: no-repeat;
		background-size: 100px auto;

		position: absolute;
		left: 50%;
		top: 50%;
		transform: translate(-50%, -50%);
	}
}
