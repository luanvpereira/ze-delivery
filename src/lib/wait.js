const ids = {};

const wait = (id, time) => {
	return new Promise(resolve => {
		clearTimeout(ids[id]);
		ids[id] = setTimeout(() => {
			resolve();
		}, time);
	});
};

export default wait;
