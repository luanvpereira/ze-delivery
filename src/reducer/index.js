import { createStore, combineReducers, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';

export const middlewares = () => [thunk];

export const initializeStore = (initialState = {}) =>
	createStore(
		combineReducers({}),
		initialState,
		composeWithDevTools(applyMiddleware(...middlewares())),
	);
