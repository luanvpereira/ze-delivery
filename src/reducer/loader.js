import {
	SHOW_LOADER,
	HIDE_LOADER
} from '../types';

export default (state = false, { type }) => {
	switch (type) {
		case SHOW_LOADER:
			return true;

		case HIDE_LOADER:
			return false;

		default:
			return state;
	}
};
