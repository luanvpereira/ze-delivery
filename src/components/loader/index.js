import React from 'react';

import style from './style.scss';

const Loader = () => (
	<div className={style.loader}></div>
);

export default Loader;
