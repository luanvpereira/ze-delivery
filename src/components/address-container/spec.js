jest.mock('react-apollo', () => ({
	withApollo: Component => Component
}));

jest.mock('../../actions/address');

jest.mock('../../lib/wait', () => jest.fn().mockImplementation(() => Promise.resolve(true)));

jest.mock('next/router', () => ({
	push: jest.fn().mockImplementation(() => Promise.resolve([]))
}));

jest.mock('../../actions/loader');

import React from 'react';
import Router from 'next/router';
import { Provider } from 'react-redux';
import { mount, shallow } from 'enzyme';

import AddressContainer, { PureAddressContainer, mapDispatchToProps } from '../address-container';
import validAddressReponse from '../../../test-utils/mocked-store/address';
import wait from '../../lib/wait';

import * as loaderActions from '../../actions/loader';

import { initializeStore } from '../../reducer';

const defaultStoreValues = {};

const pocMethodSearch = {
	data: {
		pocSearch: [
			{
				id: Math.floor(Math.random() * 10)
			}
		]
	}
};

const invalidPocMethodSearch = {
	data: {
		pocSearch: []
	}
};

const clientPocMethodSearchSuccessResponse = {
	client: {
		query: jest.fn().mockImplementation(() => Promise.resolve(pocMethodSearch))
	}
};


const clientPocMethodSearchErrorResponse = {
	client: {
		query: jest.fn().mockImplementation(() => Promise.resolve(invalidPocMethodSearch))
	}
};

const getComponent = ({
	store = defaultStoreValues,
	props = {}
} = {}) => mount(
	<Provider store={initializeStore(store)}>
		<AddressContainer {...props} />
	</Provider>).find('AddressContainer');

describe('<AddressContainer/>', () => {
	describe('#handleChange', () => {
		it('should get addressess when pass a valid address', async () => {
			const component = getComponent();
			const instance = component.instance();

			const isTyping = jest.fn();

			Object.defineProperty(instance, 'isTyping', {
				set: value => {
					isTyping(value);
					return value;
				}
			});

			const event = {
				target: {
					value: 'rua cuiaba'
				}
			};

			await instance.handleChange(event);
			expect(wait).toHaveBeenCalledWith('TYPING', 600);
			expect(component.state().addressess).toEqual(validAddressReponse);
			expect(isTyping.mock.calls[0][0]).toBe(true);
			expect(isTyping.mock.calls[1][0]).toBe(false);
		});

		it('should reset address list when cannot get addresss', async () => {
			const resetAddress = jest.spyOn(PureAddressContainer.prototype, 'resetAddressess');
			const component = getComponent();

			const event = {
				target: {
					value: ''
				}
			};

			await component.instance().handleChange(event);

			expect(resetAddress).toHaveBeenCalled();
		});
	});

	describe('#goToProducts', () => {
		it('should get a first `method` and redirect to /products page', async () => {
			const component = getComponent({
				props: {
					...clientPocMethodSearchSuccessResponse
				}
			});

			const { id } = pocMethodSearch.data.pocSearch[0];

			await component.instance().goToProducts();

			expect(Router.push).toHaveBeenCalledWith(`/products/${id}`);
		});

		it('shoud call #showLoader when consult pocSearchMethod', async () => {
			const component = getComponent({
				props: {
					...clientPocMethodSearchSuccessResponse
				}
			});

			await component.instance().goToProducts();

			expect(loaderActions.showLoader).toHaveBeenCalled();
		});

		it('shoud call #hideLoader when pocSearchMethod does not return anything', async () => {
			const component = getComponent({
				props: {
					...clientPocMethodSearchErrorResponse
				}
			});

			await component.instance().goToProducts();

			expect(loaderActions.hideLoader).toHaveBeenCalled();
		});
	});

	describe('#handleClick', () => {
		it('should call #goToProducts with lat e long', async () => {
			const goToProductsSpy = jest
				.spyOn(PureAddressContainer.prototype, 'goToProducts')
				.mockImplementation(() => Promise.resolve([]));

			const component = getComponent();

			const index = 1;


			const selectedLocation = validAddressReponse[index];

			const {
				geometry: {
					location: { lat, lng }
				}
			} = selectedLocation;

			component.setState({
				addressess: validAddressReponse
			});

			await component.instance().handleClick(index)();

			expect(goToProductsSpy).toHaveBeenCalledWith(String(lat), String(lng));
		});
	});

	describe('listeners', () => {
		it('should call #handleChange when change input value', () => {
			const handleChangeSpy = jest.spyOn(PureAddressContainer.prototype, 'handleChange');
			const component = getComponent();

			component.find('input').simulate('change');

			expect(handleChangeSpy).toHaveBeenCalled();
		});

		it('should call #handleClick when click an addresss item', () => {
			const handleClickSpy = jest.spyOn(PureAddressContainer.prototype, 'handleClick');
			const props = mapDispatchToProps(() => { });
			const component = mount(<PureAddressContainer {...props} />);
			const index = 1;

			component.setState({
				addressess: validAddressReponse
			});

			component.find('li').at(index).simulate('click');

			expect(handleClickSpy).toHaveBeenCalledWith(index);
		});
	});

	describe('#render', () => {
		it('should render an <AddressContainer/>', () => {
			const props = mapDispatchToProps(() => { });
			const component = shallow(<PureAddressContainer {...props} />);

			component.setState({
				addressess: validAddressReponse
			});

			expect(component).toMatchSnapshot();
		});
	});
});
