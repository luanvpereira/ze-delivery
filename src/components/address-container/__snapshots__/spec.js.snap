// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddressContainer/> #render should render an <AddressContainer/> 1`] = `
<Fragment>
  <input
    onChange={[Function]}
    type="text"
  />
  <ul>
    <li
      key="ChIJn826YLJezpQRDhdmNWzKsdw"
      onClick={[Function]}
    >
      R. Cuiabá - Alto da Mooca, São Paulo - SP, Brazil
    </li>
    <li
      key="ChIJF9lBHVcFbJIRt9jT0OF5WWs"
      onClick={[Function]}
    >
      R. Cuiabá - Nossa Sra. das Gracas, Manaus - AM, 69053-120, Brazil
    </li>
  </ul>
</Fragment>
`;
